### PR TITLE
Better error for failed wildcard path in `dd.read_csv()` (fixes #8878)

### LINKED ISSUE #8878: Failed wildcard matching with dd.read_csv("s3://[...]/*.csv", compression='infer') raises unintuitive error message
**What happened**:

A wildcard match that matches no S3 objects in `dd.read_csv(compression='infer')` will return an unintuitive IndexError.

```
IndexError: list index out of range
```

**What you expected to happen**:

That wildcard match should likely return the following error message instead:

```
OSError: s3://private_bucket/path/to/files/TYPO/data_*.csv resolved to no files
```

**Minimal Complete Verifiable Example**:

Paths redacted as this is a private S3 bucket, if someone has a public bucket I could try and provide a fully reproducible example.

I tried a wildcard match on a bunch of files in S3, but misspecified the path/key for a wildcard match:

```
import dask.dataframe as dd
data = dd.read_csv("s3://private_bucket/path/to/files/TYPO/data_*.csv", include_path_column=False, assume_missing=True)
```
I got this error:

```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Input In [11], in <cell line: 1>()
----> 1 data = dd.read_csv("s3://private_bucket/path/to/files/TYPO/data_*.csv",include_path_column=False,assume_missing=True)

File ~/miniconda3/lib/python3.9/site-packages/dask/dataframe/io/csv.py:742, in make_reader.<locals>.read(urlpath, blocksize, lineterminator, compression, sample, sample_rows, enforce, assume_missing, storage_options, include_path_column, **kwargs)
    729 def read(
    730     urlpath,
    731     blocksize="default",
   (...)
    740     **kwargs,
    741 ):
--> 742     return read_pandas(
    743         reader,
    744         urlpath,
    745         blocksize=blocksize,
    746         lineterminator=lineterminator,
    747         compression=compression,
    748         sample=sample,
    749         sample_rows=sample_rows,
    750         enforce=enforce,
    751         assume_missing=assume_missing,
    752         storage_options=storage_options,
    753         include_path_column=include_path_column,
    754         **kwargs,
    755     )

File ~/miniconda3/lib/python3.9/site-packages/dask/dataframe/io/csv.py:521, in read_pandas(reader, urlpath, blocksize, lineterminator, compression, sample, sample_rows, enforce, assume_missing, storage_options, include_path_column, **kwargs)
    516     paths = get_fs_token_paths(urlpath, mode="rb", storage_options=storage_options)[
    517         2
    518     ]
    520     # Infer compression from first path
--> 521     compression = infer_compression(paths[0])
    523 if blocksize == "default":
    524     blocksize = AUTO_BLOCKSIZE

IndexError: list index out of range
```

`paths` likely returns an empty list, and Python raises an error.

But if I add `compression=None` to stop compression inference:

```
data = dd.read_csv("s3://private_bucket/path/to/files/TYPO/data_*.csv",include_path_column=False,assume_missing=True,compression=None)
```

I instead get this much more intuitive error:

```
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
Input In [12], in <cell line: 1>()
----> 1 data = dd.read_csv("s3://private_bucket/path/to/files/TYPO/data_*.csv",include_path_column=False,assume_missing=True,compression=None)

File ~/miniconda3/lib/python3.9/site-packages/dask/dataframe/io/csv.py:742, in make_reader.<locals>.read(urlpath, blocksize, lineterminator, compression, sample, sample_rows, enforce, assume_missing, storage_options, include_path_column, **kwargs)
    729 def read(
    730     urlpath,
    731     blocksize="default",
   (...)
    740     **kwargs,
    741 ):
--> 742     return read_pandas(
    743         reader,
    744         urlpath,
    745         blocksize=blocksize,
    746         lineterminator=lineterminator,
    747         compression=compression,
    748         sample=sample,
    749         sample_rows=sample_rows,
    750         enforce=enforce,
    751         assume_missing=assume_missing,
    752         storage_options=storage_options,
    753         include_path_column=include_path_column,
    754         **kwargs,
    755     )

File ~/miniconda3/lib/python3.9/site-packages/dask/dataframe/io/csv.py:546, in read_pandas(reader, urlpath, blocksize, lineterminator, compression, sample, sample_rows, enforce, assume_missing, storage_options, include_path_column, **kwargs)
    544     sample = blocksize
    545 b_lineterminator = lineterminator.encode()
--> 546 b_out = read_bytes(
    547     urlpath,
    548     delimiter=b_lineterminator,
    549     blocksize=blocksize,
    550     sample=sample,
    551     compression=compression,
    552     include_path=include_path_column,
    553     **(storage_options or {}),
    554 )
    556 if include_path_column:
    557     b_sample, values, paths = b_out

File ~/miniconda3/lib/python3.9/site-packages/dask/bytes/core.py:84, in read_bytes(urlpath, delimiter, not_zero, blocksize, sample, compression, include_path, **kwargs)
     81 fs, fs_token, paths = get_fs_token_paths(urlpath, mode="rb", storage_options=kwargs)
     83 if len(paths) == 0:
---> 84     raise OSError("%s resolved to no files" % urlpath)
     86 if blocksize is not None:
     87     if isinstance(blocksize, str):

OSError: s3://private_bucket/path/to/files/TYPO/data_*.csv resolved to no files
```

**Anything else we need to know?**:

**Environment**:

- Dask version: 2022.02.1
- Python version: 3.9
- Operating System: Ubuntu 20.04.3 LTS
- Install method (conda, pip, source): conda

### PR DESCRIPTION
**Pull request by rfilmyer**
Makes an empty wildcard path URL specification in `dd.read_csv(compression='infer')` fail with `OSError: [urlpath] resolved to no files` instead of `IndexError: list index out of range`

- [X] Closes #8878
- [X] Tests added / passed
- [X] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

File: dask/dataframe/io/csv.py [hunk 1]
```diff
@@ -517,6 +517,10 @@ def read_pandas(
             2
         ]
 
+        # Check for at least one valid path
+        if len(paths) == 0:
+            raise OSError("%s resolved to no files" % urlpath)
```
Comment (pavithraes) [inline dask/dataframe/io/csv.py:522]: I just realized that we'd like to have Python-3 optimization wherever we can in the codebase, so we may want to use f-strings here? It's just a nice-to-have, so completely up to you. :)

**Comment by pavithraes** [comment]
@rfilmyer Thanks for opening this PR, it looks good to me!

File: dask/dataframe/io/csv.py (hunk 1 above)
Comment (rfilmyer) [reply dask/dataframe/io/csv.py:522]: I honestly use f-strings way more, I'm definitely fine to change this. Wanted to keep consistency with existing code, but if we're trying to use f-strings more I'll change it.

**Comment by pavithraes** [comment]
Thanks for the updates, @rfilmyer! The failing windows test seems unrelated. :)

@jsignell I think we're good to go! Could you please take a quick look and help merge?

**Pull request merged**